### ISSUE TITLE
False positive for "Argument missing for parameter"

### ISSUE THREAD

**Issue by last-partizan** [microsoft/pyright#2647]
**Describe the bug**
When using decorator in the following way, there is false positive "Argument missing for parameter"

**To Reproduce**
Create file and type-check it.

**Expected behavior**
There should be no error "Argument missing for parameter"

**Screenshots or Code**

```python
def register(func=None):
    def decorator(func):

        def new_func(arg1, arg2):
            ...
        return new_func
    return decorator if not func else decorator(func)


@register()
def function():
    ...
```

```
pyright ~/src/pyright-test/decorator.py
...
/home/serg/src/pyright-test/decorator.py
  /home/serg/src/pyright-test/decorator.py:10:2 - error: Argument missing for parameter "arg2" (reportGeneralTypeIssues)1 error, 0 warnings, 0 infos
Completed in 0.532sec
```

**Comment by erictraut** [comment]
The function `register` returns a union of two types: 
`(func: Unknown) -> ((arg1: Unknown, arg2: Unknown) -> None)`
`(arg1: Unknown, arg2: Unknown) -> None`

To properly type annotate this decorator, you will need to use an overload to specify that it returns one of these two types when the input argument is `None` and the other when it's not `None`.

```python
@overload
def register(func: None = None) -> Callable[..., Callable[[Any, Any], None]]:
    ...

@overload
def register(func: Callable[..., Any]) -> Callable[[Any, Any], None]:
    ...
```
